src/plugins/intel_cpu/src/nodes/kernels/registers_pool.hpp

```diff
@@ -0,0 +1,322 @@
+                          std::is_same<TReg, Xbyak::Reg8>::value || std::is_same<TReg, Xbyak::Reg16>::value ||
+                          std::is_same<TReg, Xbyak::Reg32>::value || std::is_same<TReg, Xbyak::Reg64>::value ||
+                          std::is_same<TReg, Xbyak::Opmask>::value,
+                          "The type is not supported for the RegistersPool::Reg template");
+        }
+        Reg(const RegistersPool::Ptr& regPool) { initialize(regPool); }
+        Reg(const RegistersPool::Ptr& regPool, int requestedIdx) { initialize(regPool, requestedIdx); }
+        ~Reg() { release(); }
+        Reg& operator=(Reg&& other)  noexcept {
+            release();
+            reg = other.reg;
+            regPool = std::move(other.regPool);
+            return *this;
+        }
+        Reg(Reg&& other)  noexcept : reg(other.reg), regPool(std::move(other.regPool)) {}
+        operator TReg&() { ensureValid(); return reg; }
+        operator const TReg&() const { ensureValid(); return reg; }
+        operator Xbyak::RegExp() const { ensureValid(); return reg; }
+        int getIdx() const { ensureValid(); return reg.getIdx(); }
+        friend Xbyak::RegExp operator+(const Reg& lhs, const Xbyak::RegExp& rhs) {
+            lhs.ensureValid();
+            return lhs.operator Xbyak::RegExp() + rhs;
+        }
+        void release() {
+            if (regPool) {
+                regPool->returnToPool(reg);
+                regPool.reset();
+            }
+        }
+        bool isInitialized() const { return static_cast<bool>(regPool); }
+
+    private:
+        void ensureValid() const {
+            if (!isInitialized()) {
+                IE_THROW() << "RegistersPool::Reg is either not initialized or released";
+            }
+        }
+
+        void initialize(const RegistersPool::Ptr& pool, int requestedIdx = anyIdx) {
+            release();
+            reg = TReg(pool->template getFree<TReg>(requestedIdx));
+            regPool = pool;
+        }
+
+    private:
+        TReg reg;
+        RegistersPool::Ptr regPool;
+    };
+
+    virtual ~RegistersPool() {
+        checkUniqueAndUpdate(false);
+    }
+
+    template <x64::cpu_isa_t isa>
+    static Ptr create(std::initializer_list<Xbyak::Reg> regsToExclude);
+
+    template<typename TReg>
+    size_t countFree() const {
+        if (std::is_same<TReg, Xbyak::Xmm>::value || std::is_same<TReg, Xbyak::Ymm>::value || std::is_same<TReg, Xbyak::Zmm>::value) {
+            return simdSet.countUnused();
+        } else if (std::is_same<TReg, Xbyak::Reg8>::value || std::is_same<TReg, Xbyak::Reg16>::value ||
+                   std::is_same<TReg, Xbyak::Reg32>::value || std::is_same<TReg, Xbyak::Reg64>::value) {
+            return generalSet.countUnused();
+        } else if (std::is_same<TReg, Xbyak::Opmask>::value) {
+            return countUnusedOpmask();
+        }
```

Comment:
For me it is subjective either to use `std::enable_if` or `std::is_same`, because `std::enable_if` is usually decrease readability.
Also this code with `std::is_same` could be improved in future when C++17 will be add for OpenVINO with `if constexpr`

Reply:
Let's look at pros and cons of using SFINAE here:
Pros:
1. We remove conditional jumps in the runtime. Still, I'm not sure that this is a good idea to check compile time params at runtime.
2. We introduce early error check at the compile time, when by some reason a wrong template parameter is used.
3. We generate less code, since for all different register types the full if checks list will not be generated.

Cons:
1. Less readable code (subjective). 

Also those functions are not used directly in the normal register pool usage flow, so m.b. it dose make sense to switch to `enable_if`? Personally I have no idea when we move to C++17, so I would not consider it as a good argument to make decisions right now.

Reply:
Thanks for the hint. I changed to use SFINAE for the countFree, getFree, and returnToPool methods.

Reply:
@nshchego 
> Let's look at pros and cons of using SFINAE here: Pros:
> 
>     1. We remove conditional jumps in the runtime. Still, I'm not sure that this is a good idea to check compile time params at runtime.

Actually in Release mode this if statement would be constexpresion and removed by optimizer since it is would be:
```cpp
if (true) {
  ...
} else {
  ...
}
```
This else case and other if-s would be removed in Release

> 
>     2. We introduce early error check at the compile time, when by some reason a wrong template parameter is used.

Same could be done using `static_assert` without introducing a new functions

> 
>     3. We generate less code, since for all different register types the full `if` checks list will not be generated.

To reduce number of of could should be removed template, but enable_if do not really change number of generated code.
Instead of one big function, it will generate lots of smaller one